repo/fsrepo/datastores.go

```diff
@@ -0,0 +1,126 @@
+package fsrepo
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+
+	repo "github.com/ipfs/go-ipfs/repo"
+
+	measure "gx/ipfs/QmNPv1yzXBqxzqjfTzHCeBoicxxZgHzLezdY2hMCZ3r6EU/go-ds-measure"
+	flatfs "gx/ipfs/QmXZEfbEv9sXG9JnLoMNhREDMDgkq5Jd7uWJ7d77VJ4pxn/go-ds-flatfs"
+
+	ds "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore"
+	mount "gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/syncmount"
+
+	levelds "gx/ipfs/QmaHHmfEozrrotyhyN44omJouyuEtx6ahddqV6W5yRaUSQ/go-ds-leveldb"
+	ldbopts "gx/ipfs/QmbBhyDKsY4mbY6xsKt3qu9Y7FPvMJ6qbD8AMjYYvPRw1g/goleveldb/leveldb/opt"
+)
+
+func (r *FSRepo) constructDatastore(params map[string]interface{}) (repo.Datastore, error) {
+	switch params["type"] {
+	case "mount":
+		mounts, ok := params["mounts"].([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("mounts field wasnt an array")
+		}
+
+		return r.openMountDatastore(mounts)
+	case "flatfs":
+		return r.openFlatfsDatastore(params)
+	case "mem":
+		return ds.NewMapDatastore(), nil
+	case "log":
+		child, err := r.constructDatastore(params["child"].(map[string]interface{}))
```

Comment:
probably want to validate the existence of the child key, and make sure its type asserted properly